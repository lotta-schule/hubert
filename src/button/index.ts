export * from './BaseButton';
export * from './Button';
export * from './ButtonGroup';
export * from './NavigationButton';
export * from './ScrollToTopButton';
