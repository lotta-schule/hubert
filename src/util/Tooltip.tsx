import * as React from 'react';
import { useTooltip, useTooltipTrigger } from '@react-aria/tooltip';
import {
  useTooltipTriggerState,
  TooltipTriggerState,
} from '@react-stately/tooltip';
import { mergeProps } from '@react-aria/utils';
import clsx from 'clsx';

import styles from './Tooltip.module.scss';

export interface TooltipProps extends React.HTMLAttributes<HTMLDivElement> {
  children: React.ReactElement;
  label?: string | React.ReactElement<{}>;
}

export const Tooltip = React.forwardRef(
  (
    { children, className, label, ...props }: TooltipProps,
    refProp: React.ForwardedRef<HTMLDivElement | null>
  ) => {
    const state = useTooltipTriggerState({});
    const ref = React.useRef<HTMLDivElement | null>(
      (refProp as React.MutableRefObject<HTMLDivElement>)?.current ?? null
    );

    let { triggerProps, tooltipProps } = useTooltipTrigger({}, state, ref);

    return (
      <div className={styles.root}>
        {React.cloneElement(children, { ref, ...triggerProps })}
        <TooltipOverlay
          className={className}
          state={state}
          label={label ?? children}
          {...mergeProps(props, tooltipProps)}
        />
      </div>
    );
  }
);
Tooltip.displayName = 'Tooltip';

export interface TooltipOverlayProps
  extends React.HTMLAttributes<HTMLDivElement> {
  label?: string | React.ReactElement<{}>;
  state: TooltipTriggerState;
}

export const TooltipOverlay = React.memo(
  React.forwardRef(
    (
      { state, label, className, ...props }: TooltipOverlayProps,
      ref: React.ForwardedRef<HTMLDivElement>
    ) => {
      const { tooltipProps } = useTooltip(props, state);
      return state.isOpen ? (
        <div
          className={clsx(className, styles.tooltip)}
          ref={ref}
          {...mergeProps(props, tooltipProps)}
        >
          {label}
        </div>
      ) : null;
    }
  )
);
TooltipOverlay.displayName = 'TooltipOverlay';
