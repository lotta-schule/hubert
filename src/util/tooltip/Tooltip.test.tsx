import { render } from '@testing-library/react';
import { Button } from '../../button/Button';
import { Tooltip } from './Tooltip';
import userEvent from '@testing-library/user-event';

describe('util/Tooltip', () => {
  it('should render', () => {
    const screen = render(
      <Tooltip label="Test">
        <Button>Test</Button>
      </Tooltip>
    );
    expect(screen.container).toMatchSnapshot();
  });

  it('should keep a button clickable', () => {
    const onClick = jest.fn();
    const screen = render(
      <Tooltip label="Test">
        <Button onClick={onClick}>Test</Button>
      </Tooltip>
    );
    userEvent.click(screen.getByRole('button'));
    expect(onClick).toHaveBeenCalled();
  });
});
