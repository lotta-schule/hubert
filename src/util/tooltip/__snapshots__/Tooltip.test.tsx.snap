// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`util/Tooltip should render 1`] = `
<div>
  <div
    class="root"
  >
    <div
      class="trigger"
    >
      <button
        aria-current="false"
        class="lotta-base-button variant__default lotta-button"
        role="button"
        type="button"
      >
        <span
          class="lotta-button_text"
        >
          Test
        </span>
      </button>
    </div>
  </div>
</div>
`;
